sklearn/calibration.py

```diff
@@ -562,9 +567,26 @@ def calibration_curve(y_true, y_prob, normalize=False, n_bins=5):
     bins = np.linspace(0., 1. + 1e-8, n_bins + 1)
     binids = np.digitize(y_prob, bins) - 1
 
-    bin_sums = np.bincount(binids, weights=y_prob, minlength=len(bins))
-    bin_true = np.bincount(binids, weights=y_true, minlength=len(bins))
-    bin_total = np.bincount(binids, minlength=len(bins))
+    if sample_weight is None:
+        bin_sums = np.bincount(binids, weights=y_prob, minlength=len(bins))
+        bin_true = np.bincount(binids, weights=y_true, minlength=len(bins))
+        bin_total = np.bincount(binids, minlength=len(bins))
+    else:
+        sample_weight = check_array(sample_weight, ensure_2d=False)
+        check_consistent_length(y_true, sample_weight)
+
+        # Check that the sample weights sum is positive
+        if sample_weight.sum() <= 0:
```

Comment:
does it make sense to have any negative weight?

Reply:
I agree, we should throw an error as soon as any sample_weight is negative. Wondering about zeros in sample weights though...

Reply:
Is there a specific reason why an individual sample weight may not be negative? Can a paper be cited that states negative weights are statistically unsound? The paper I linked above shows that they are situations in which they are strictly necessary. I hope this issue can be reconsidered.

Reply:
#3774 for reference to an old discussion around negative sample weights.